File: desktop/src/renderer/lib/task-events.ts
export type TaskType = "collection" | "scrape" | "calibration";

export type TaskPrompt = {
  step: string;
  kind: "position" | "ack" | "integer" | "confirm";
  title: string;
  message: string;
  default_value?: number;
  min_value?: number;
  confirm_label?: string;
  reject_label?: string;
};

export type TaskStartedEvent = {
  type: "started";
  task_id: string;
  task_type: TaskType;
};

export type TaskLogEvent = {
  type: "log";
  task_id: string;
  message: string;
};

export type TaskProgressEvent = {
  type: "progress";
  task_id: string;
  current: number;
  total: number;
  message: string;
  success?: number;
  failed?: number;
};

export type TaskStatusEvent = {
  type: "status";
  task_id: string;
  status: string;
  message: string;
};

export type TaskPromptEvent = {
  type: "prompt";
  task_id: string;
  prompt: TaskPrompt;
};

export type TaskCompletedEvent = {
  type: "completed";
  task_id: string;
  task_type: TaskType;
};

export type TaskErrorEvent = {
  type: "error";
  task_id: string;
  message: string;
};

export type TaskStoppedEvent = {
  type: "stopped";
  task_id: string;
  reason: string;
};

export type TaskCancelledEvent = {
  type: "cancelled";
  task_id: string;
  reason: string;
};

export type TaskEvent =
  | TaskStartedEvent
  | TaskLogEvent
  | TaskProgressEvent
  | TaskStatusEvent
  | TaskPromptEvent
  | TaskCompletedEvent
  | TaskErrorEvent
  | TaskStoppedEvent
  | TaskCancelledEvent;

export type TaskSnapshotPayload = {
  task_id: string;
  task_type: TaskType;
  active: boolean;
  stopping: boolean;
  prompt?: TaskPrompt | null;
  events: TaskEvent[];
};

export type TaskProgressSummary = {
  current: number;
  total: number;
  message: string;
  percent: number;
  hasProgress: boolean;
  success: number | null;
  failed: number | null;
};

export type TaskSessionSummary = {
  title: string;
  description: string;
};

export function getTaskDisplayName(taskType: TaskType) {
  if (taskType === "collection") {
    return "采集";
  }

  if (taskType === "scrape") {
    return "抓取";
  }

  return "校准";
}

export function isTerminalTaskEvent(event: TaskEvent) {
  return event.type === "completed" || event.type === "error" || event.type === "stopped" || event.type === "cancelled";
}

export function formatTaskEventMessage(event: TaskEvent) {
  switch (event.type) {
    case "started":
      return `${getTaskDisplayName(event.task_type)}任务已开始`;
    case "log":
      return event.message;
    case "progress":
      return event.message || `已处理 ${event.current}/${event.total}`;
    case "status":
      return event.message || event.status;
    case "prompt":
      return event.prompt.message;
    case "completed":
      return `${getTaskDisplayName(event.task_type)}任务已完成`;
    case "error":
      return event.message || "任务失败";
    case "stopped":
      return event.reason || "任务已停止";
    case "cancelled":
      return event.reason || "任务已取消";
  }
}

export function summarizeTaskProgress(events: TaskEvent[]): TaskProgressSummary {
  const lastProgressEvent = [...events].reverse().find((event): event is TaskProgressEvent => event.type === "progress");

  if (!lastProgressEvent) {
    return {
      current: 0,
      total: 0,
      message: "",
      percent: 0,
      hasProgress: false,
      success: null,
      failed: null,
    };
  }

  const safeTotal = Math.max(lastProgressEvent.total, 0);
  const safeCurrent = Math.max(lastProgressEvent.current, 0);
  const percent = safeTotal > 0 ? Math.min(Math.round((safeCurrent / safeTotal) * 100), 100) : 0;

  return {
    current: safeCurrent,
    total: safeTotal,
    message: lastProgressEvent.message,
    percent,
    hasProgress: true,
    success: typeof lastProgressEvent.success === "number" ? lastProgressEvent.success : null,
    failed: typeof lastProgressEvent.failed === "number" ? lastProgressEvent.failed : null,
  };
}

function taskEventsMatch(left: TaskEvent, right: TaskEvent) {
  return JSON.stringify(left) === JSON.stringify(right);
}

export function mergeTaskEvents(existing: TaskEvent[], snapshotEvents: TaskEvent[]) {
  if (existing.length === 0) {
    return [...snapshotEvents];
  }

  if (snapshotEvents.length < existing.length) {
    return [...snapshotEvents];
  }

  const matchesPrefix = existing.every((event, index) => taskEventsMatch(event, snapshotEvents[index]));
  if (!matchesPrefix) {
    return [...snapshotEvents];
  }

  return existing.concat(snapshotEvents.slice(existing.length));
}

export function summarizeTaskSession(
  snapshot: TaskSnapshotPayload | null,
  options?: {
    isStarting?: boolean;
    isStopping?: boolean;
  },
): TaskSessionSummary {
  if (options?.isStarting) {
    return {
      title: "启动中",
      description: "正在创建任务，请稍候。",
    };
  }

  if (!snapshot) {
    return {
      title: "未开始",
      description: "点击开始按钮后，任务日志和进度会在这里更新。",
    };
  }

  const lastTerminalEvent = [...snapshot.events].reverse().find((event) => isTerminalTaskEvent(event));
  if (lastTerminalEvent?.type === "completed") {
    return {
      title: "已完成",
      description: "任务已经完成，可以重新开始新的任务。",
    };
  }

  if (lastTerminalEvent?.type === "error") {
    return {
      title: "失败",
      description: lastTerminalEvent.message || "任务执行失败。",
    };
  }

  if (lastTerminalEvent?.type === "stopped" || lastTerminalEvent?.type === "cancelled") {
    return {
      title: "已停止",
      description: lastTerminalEvent.reason || "任务已停止。",
    };
  }

  if (options?.isStopping || snapshot?.stopping) {
    return {
      title: "停止中",
      description: "停止请求已发送，正在等待任务结束。",
    };
  }

  return {
    title: snapshot.active ? "进行中" : "已结束",
    description: snapshot.active
      ? "任务正在运行，页面会自动轮询最新事件。"
      : "任务已经结束，当前显示的是最后一次快照。",
  };
}

File: desktop_backend/task_events.py
from __future__ import annotations

from typing import TypedDict


class TaskPrompt(TypedDict, total=False):
    step: str
    kind: str
    title: str
    message: str
    default_value: int
    min_value: int
    confirm_label: str
    reject_label: str


class TaskEvent(TypedDict, total=False):
    type: str
    task_id: str
    task_type: str
    message: str
    status: str
    reason: str
    current: int
    total: int
    success: int
    failed: int
    prompt: TaskPrompt


def _normalize_text(value: object) -> str:
    return "" if value is None else str(value)


def _normalize_int(value: object) -> int:
    return int(value)


def build_started_event(*, task_id: str, task_type: str) -> TaskEvent:
    return {
        "type": "started",
        "task_id": _normalize_text(task_id),
        "task_type": _normalize_text(task_type),
    }


def build_log_event(*, task_id: str, message: object) -> TaskEvent:
    return {
        "type": "log",
        "task_id": _normalize_text(task_id),
        "message": _normalize_text(message),
    }


def build_progress_event(
    *,
    task_id: str,
    current: object,
    total: object,
    message: object = "",
    success: object | None = None,
    failed: object | None = None,
) -> TaskEvent:
    event: TaskEvent = {
        "type": "progress",
        "task_id": _normalize_text(task_id),
        "current": _normalize_int(current),
        "total": _normalize_int(total),
        "message": _normalize_text(message),
    }
    if success is not None:
        event["success"] = _normalize_int(success)
    if failed is not None:
        event["failed"] = _normalize_int(failed)
    return event


def build_status_event(*, task_id: str, status: object, message: object = "") -> TaskEvent:
    return {
        "type": "status",
        "task_id": _normalize_text(task_id),
        "status": _normalize_text(status),
        "message": _normalize_text(message),
    }


def build_prompt_event(*, task_id: str, prompt: TaskPrompt) -> TaskEvent:
    normalized_prompt: TaskPrompt = {
        "step": _normalize_text(prompt.get("step")),
        "kind": _normalize_text(prompt.get("kind")),
        "title": _normalize_text(prompt.get("title")),
        "message": _normalize_text(prompt.get("message")),
    }
    if "default_value" in prompt and prompt["default_value"] is not None:
        normalized_prompt["default_value"] = _normalize_int(prompt["default_value"])
    if "min_value" in prompt and prompt["min_value"] is not None:
        normalized_prompt["min_value"] = _normalize_int(prompt["min_value"])
    if "confirm_label" in prompt:
        normalized_prompt["confirm_label"] = _normalize_text(prompt["confirm_label"])
    if "reject_label" in prompt:
        normalized_prompt["reject_label"] = _normalize_text(prompt["reject_label"])

    return {
        "type": "prompt",
        "task_id": _normalize_text(task_id),
        "prompt": normalized_prompt,
    }


def build_completed_event(*, task_id: str, task_type: str) -> TaskEvent:
    return {
        "type": "completed",
        "task_id": _normalize_text(task_id),
        "task_type": _normalize_text(task_type),
    }


def build_error_event(*, task_id: str, message: object) -> TaskEvent:
    return {
        "type": "error",
        "task_id": _normalize_text(task_id),
        "message": _normalize_text(message),
    }


def build_stopped_event(*, task_id: str, reason: object = "") -> TaskEvent:
    return {
        "type": "stopped",
        "task_id": _normalize_text(task_id),
        "reason": _normalize_text(reason),
    }


def build_cancelled_event(*, task_id: str, reason: object = "") -> TaskEvent:
    return {
        "type": "cancelled",
        "task_id": _normalize_text(task_id),
        "reason": _normalize_text(reason),
    }
